subsys/net/ip/net_core.c

```diff
@@ -123,35 +121,48 @@ static inline enum net_verdict process_data(struct net_pkt *pkt,
 		}
 	}
 
-	uint8_t family = net_pkt_family(pkt);
-
-	if (IS_ENABLED(CONFIG_NET_IP) && (family == AF_INET || family == AF_INET6 ||
-					  family == AF_UNSPEC || family == AF_PACKET)) {
-		/* IP version and header length. */
-		uint8_t vtc_vhl = NET_IPV6_HDR(pkt)->vtc & 0xf0;
+	STRUCT_SECTION_FOREACH(net_l3_register, l3) {
+		if (l3->ptype != net_pkt_ll_proto_type(pkt) ||
+		    (l3->l2 != NULL &&
+		     l3->l2 != (net_pkt_iface(pkt)->if_dev->l2 == &NET_L2_GET_NAME(VIRTUAL)
```

Comment:
This needs rewriting as l2 virtual might not be required. Should be sufficient to use orig_iface always.

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/8661c3caea02990daeeb06b6e74c7f96c2fc44f3/include/zephyr/net/net_pkt.h#L408
and
https://github.com/zephyrproject-rtos/zephyr/blob/8661c3caea02990daeeb06b6e74c7f96c2fc44f3/include/zephyr/net/net_pkt.h#L418
need to be changed as well.